src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,170 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String MESSAGE_INDEX_ERROR = "Index outside the size of the list";
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private class Node<T> {
+        private T elementValue;
+        private Node<T> previous;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T elementValue, Node<T> next) {
+            this.previous = previous;
+            this.elementValue = elementValue;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkPositionForAdd(index);
+        if (index == size) {
+            addToTail(value);
+        } else {
+            insertBefore(value, node(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            addToTail(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionForGetAndSet(index);
+        return node(index).elementValue;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T remove(int index) {
+        checkPositionForGetAndSet(index);
+        return unlinkNode(node(index));
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public boolean remove(T value) {
+        if (value == null) {
+            for (Node<T> nodeForRemove = head;
+                    nodeForRemove != null; nodeForRemove = nodeForRemove.next) {
+                if (nodeForRemove.elementValue == null) {
+                    unlinkNode(nodeForRemove);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> nodeForRemove = head;
+                    nodeForRemove != null; nodeForRemove = nodeForRemove.next) {
+                if (value.equals(nodeForRemove.elementValue)) {
+                    unlinkNode(nodeForRemove);
+                    return true;
+                }
+            }
+        }
+        return false;
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public T set(T value, int index) {
+        checkPositionForGetAndSet(index);
+        Node<T> nodeForChangeValue = node(index);
+        T oldValue = nodeForChangeValue.elementValue;
+        nodeForChangeValue.elementValue = value;
+        return oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addToTail(T value) {
```

Comment:
I think it is no need to use separate method addToTail. it is the same as method add(T value) and you can just use in in other add-method.

Reply:
I also agree. But it's more convenient for me to read the code. I'll think about it.